### PR TITLE
Consider subst in prioritising placeholder inherent methods

### PR DESCRIPTION
**Pull request by HKalbasi**
Reduces unknown types on self to 18

### CODE REVIEW

File: crates/hir-ty/src/infer/expr.rs [hunk 1]
```diff
@@ -1483,10 +1483,44 @@ impl<'a> InferenceContext<'a> {
...
         );
         let (receiver_ty, method_ty, substs) = match resolved {
-            Some((adjust, func, visible)) => {
+            Some((adjust, func, visible, subst)) => {
                 let (ty, adjustments) = adjust.apply(&mut self.table, receiver_ty);
                 let generics = generics(self.db.upcast(), func.into());
-                let substs = self.substs_for_method_call(generics, generic_args);
+                let substs = match subst {
+                    Some(s) => {
+                        // FIXME: here we need the method subst, but result of `lookup_method` incorrectly returns the trait
```
Comment (Veykril) [inline crates/hir-ty/src/infer/expr.rs:1491]: We probably want to look into rustc's probing logic and copy that at some point

File: crates/hir-ty/src/tests/method_resolution.rs [hunk 2]
```diff
@@ -949,6 +949,27 @@ fn test() { S2.into(); }
...
+    }
+}
+fn test<T: Tr<i32>>(x: T) { x.tr(); }
+                          //^^^^^^ i32
+fn test<T: Tr<i32> + Tr<i64>>(x: T) { let x: i64 = x.tr(); }
+                                                 //^^^^^^ i64
+"#,
+    );
+}
+
```
Comment (Veykril) [inline crates/hir-ty/src/tests/method_resolution.rs:972]: is this basically https://github.com/rust-lang/rust-analyzer/issues/5514?

Comment (HKalbasi) [reply crates/hir-ty/src/tests/method_resolution.rs:972]: Yes. I was looking at the unknown types for finding regressions of #14470 and when I found out that this is definitely not a regression, it was too late :)

I don't think it needs chalk support, as it is a method resolution thing. The current implementation also handles the `Into2` case mentioned in the issue, I will add a test for it.

Comment (lowr) [reply crates/hir-ty/src/tests/method_resolution.rs]: I don't think this solves the general issue unfortunately -- `Into2` case that's mentioned in the issue is the following (mostly taken from the playground linked in rust-lang/chalk#584), which still fails with this PR. I don't think we can solve this without overhauling chalk, because it needs to prioritize `T: Into<i32>` bound when proving subobligation chalk produces internally which we don't have control over.

```rust
trait Into<T> {
    fn into(self) -> T;
}
impl<T> Into<T> for T {
    fn into(self) -> T { loop {} }
}
trait Into2<T> {
    fn into2(self) -> T;
}
impl<U, T> Into2<T> for U where U: Into<T> {
    fn into2(self) -> T { loop {} }
}

fn test<T: Into<i32>>(t: T) {
    let u = t.into2();
      //^i32
}
```

Comment (HKalbasi) [reply crates/hir-ty/src/tests/method_resolution.rs]: Hmm, you are right. If that issue in chalk becomes solved, our `iterate_inherent_trait_methods` will become completely unnecessary?

Comment (lowr) [reply crates/hir-ty/src/tests/method_resolution.rs]: I believe we need it regardless. That function is for prioritizing some traits (principal trait for dyn, trait bounds for placeholder) to search methods from, whereas the chalk issue is for prioritizing where clauses to be considered when solving obligations.

Consider the code below. Even if we could prove both `C: Foo` and `C: Bar` (with or without the chalk issue resolved), we need to pick `Bar::f` when resolving `v.f()`.

```rust
trait Foo {
    fn f(self);
}
trait Bar {
    fn f(self);
}

fn test<C: Bar>(v: C) {
    v.f();
}
```

Comment (HKalbasi) [reply crates/hir-ty/src/tests/method_resolution.rs]: What is the difference between placeholder bound and where clauses? In your example, when chalk issue is resolved, it should prioritize `Bar::f` since `C: Bar` is a where clause in the scope.

Comment (lowr) [reply crates/hir-ty/src/tests/method_resolution.rs]: Chalk's job is to prove/disprove stuffs and nothing more, and in the course of doing so it should prioritize bounds from where clauses (the chalk issue we're talking about). So even when it prioritizes some bounds, it goes ahead and proves `C: Foo` e.g. if there's a blanket impl `impl<T> Foo for T`. In that case `v.f()` *can* be either method, and it's *our* job to prioritize `Bar::f()` because of `C: Bar` bound.

Obviously I'm not the best explainer :sweat_smile: You might want to look at these functions and their callers in rustc: [`assemble_inherent_candidates_from_object()`][obj], [`assemble_inherent_candidates_from_param`][param]. I'd be happy to try explaining more though!

[obj]: https://github.com/rust-lang/rust/blob/660c966ff941ddf995d3251df32508b383cd4cee/compiler/rustc_hir_typeck/src/method/probe.rs#L772
[param]: https://github.com/rust-lang/rust/blob/660c966ff941ddf995d3251df32508b383cd4cee/compiler/rustc_hir_typeck/src/method/probe.rs#L821

Comment (lowr) [reply crates/hir-ty/src/tests/method_resolution.rs]: Let me rephrase what I meant: chalk prioritization should only happen when there are multiple applicable clauses to prove a goal.

- In the case of `Into2`, chalk produces a subgoal `!T: Into<?0>` (where `!T` is the placeholder from `test()`). The problem here is that it then sees two applicable clauses to solve it: `impl<T> Into<T> for T` from the blanket impl, and `T: Into<i32>` from the placeholder bound. Current chalk doesn't prioritize the latter when it should, leading to `?0` being unresolved.
- In the case of `Foo` and `Bar`, nothing prevents chalk from proving `C: Foo` and `C: Bar` because there are no conflicting clauses to prove them.

Comment (HKalbasi) [reply crates/hir-ty/src/tests/method_resolution.rs]: So, even after the chalk issue, we need something similar to this PR? Currently:
```Rust
fn f<T: Into<i32>>(t: T) {
    let x: T = t.into(); // expected T found i32
}
```
Is a compile error. But chalk with any prioritization is able to solve `T: Into<T>`, and we should prioritize `Into<i32>` in method resolution? What the current PR is doing is simple, before it when we had a `Into<i32>` trait, we prioritized `Into` methods. Now, we prioritize `Into<i32>` methods.

Comment (lowr) [reply crates/hir-ty/src/tests/method_resolution.rs]: No, chalk with prioritization won't be able to prove `T: Into<T>`, because:

1. we tell chalk to solve `T: Into<?0>`
1. chalk sees two applicable clauses: `impl<T> Into<T> for T` (blanket impl) and `T: Into<i32>` (placeholder bound)
1. chalk prioritizes the placeholder bound and unifies `?0` with `i32`
1. chalk returns unique solution with substitution `?0 = i32`

Comment (lowr) [reply crates/hir-ty/src/tests/method_resolution.rs]: Umm wait, I need to think a bit more.

Comment (HKalbasi) [reply crates/hir-ty/src/tests/method_resolution.rs]: It seems you are correct. So, in the rustc code there is no usage of `<i32>` part of the `Into<i32>` until trait solving? And we can prioritize methods of `Into<_>` and the rest will be handled by trait solver? I also tried method where bounds, but it seems they don't affect resolution, which confirms this. In the function you linked from rustc it is passing trait_refs, but it seems it later only uses the trait id?

Comment (lowr) [reply crates/hir-ty/src/tests/method_resolution.rs]: Eh, so turns out I wasn't fully correct in regards to substitution i.e. `<i32>` part. `assemble_inherent_candidates_from_param()` records trait ref as you pointed out, and its substitution seems to be used to instantiate obligations rustc feeds into its trait solver. But I don't think it matters much, because the solver can only prove `T: Into<i32>` anyway due to prioritization.

Comment (HKalbasi) [reply crates/hir-ty/src/tests/method_resolution.rs]: So, the current approach either should be completely removed, or only plays a small rule in the ideal solution. From that question, I wanted to know that does it worth to keep this as a workaround? It appears to work for the basic case of `impl Into<i32>`, but it makes our code unnecessarily ugly. So if you think the ideal solution is accessible in near future, we should wait for that and abandon this.

And if we decide to go with this as a workaround, we should check that this behaviour is a subset of the rustc's, and doesn't introduces new false positive type mismatches.

**Comment by HKalbasi** [comment]
Closing per the discussion. I still think it is an acceptable workaround for an old and high impact problem, but I see the value in complete solutions.
